docker/Dockerfile

```diff
@@ -0,0 +1,43 @@
+FROM debian:stable-slim
+
+# silence env warnings
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+    python3-minimal \
+    python3-pip \
+    python3-setuptools \
+    ca-certificates \
+    nodejs \
+    npm \
+    git 
+
+# NOTE: there is currently a compability warning with the target versions of Node vs. NPM - using latest until something is identified as the target version
+RUN rm -rf /var/lib/apt/lists/* \
+    && npm install npm@latest -g
+
+# NOTE: Python module for pip requires upgrade adhead of PIO install to work around defects
+RUN python3 -m pip install -U pip
+
+# now that pip is upgraded, we can install platformio without it crashing
+RUN python3 -m pip install platformio
+
+# Create the working directories for where to copy code for a clean build and for output
+RUN mkdir -p /espurna && \
+    mkdir -p /firmware 
+
+# Copy in source code 
+COPY . /espurna
+
+# update our working director so that ./build.sh will execute correctly
+WORKDIR /espurna/code
+
+RUN npm install --only=dev && \
+    platformio run --target clean
```

Comment:
I missed this part. We also want the libs shared? Check out the scripts/pio_pre.py, ESPURNA_PIO_SHARED_LIBRARIES env var and the platformio.ini config entry. Since PIO4 it does install libraries per-environment, so you end up with .pio/libdeps/sonoff-basic, .pio/libdeps/sonoff-pow and etc. However, since docker makes a snapshot of the espurna version and lib_deps is shared between envs, it would make sense to place libs there.

This can be a separate target for the platformio, check out https://docs.platformio.org/en/latest/projectconf/advanced_scripting.html#custom-targets
(both platform tools and libraries in the same step)

It's not on by default though (and I'll need to remember why not...) and needs that env flag to install libs there.

Reply:
Took me a bit of digging on this to try and understand what you're looking for and I'm still not certain I understand what change your asking for here (probably because I'm not as familiar with the build environment setup). I think it's as simple as calling a `platformio lib install` ... but that does significantly increase the build time of the docker image, especially if you really only wanted it to build one target so you didn't have to setup a ton of environments (a couple of minutes to 20+ minutes). My original use case for this was far more for the person that had 1 target to build as opposed to **all** of the targets to build, given that if your run with a single target, it still installs the associated libraries that you need to generate a firmware image.

Reply:
Not quite, the idea is that libs would be fetched just once into this dir:
https://github.com/xoseperez/espurna/blob/ac6f0b71e63206562acfdb8cec8e8e0331cb1f79/code/platformio.ini#L169-L170
Historically, PIO used a common library directory as .pio/libdeps/ until PIO4, which is what this essentially does try to simulate. No need to call each $env, just *something* that will fetch the libraries + build tools once.